File: java/src/main/java/bowling/Bowling.java
package bowling;

import java.util.ArrayList;

class Bowling {

    private static final int MAX_FRAMES_PER_GAME = 10;
    private ArrayList<Frame> frames;
    private Frame currentFrame;

    int computeScore(String line) {
        frames = new ArrayList<>();
        convertLineToFrames(line);

        return frames.stream()
                .filter(frame -> !frame.isBonusFrame())
                .reduce(0, (score, frame) -> score + frame.getFrameScore(), Integer::sum);
    }

    private void convertLineToFrames(String line) {
        for (String roll : line.split("")) {
            currentFrame().addRoll(roll);
        }
    }

    private Frame currentFrame() {
        if (currentFrame == null) {
            currentFrame = new Frame(false);
            frames.add(currentFrame);
        }
        if (currentFrame.isFinished()) {
            currentFrame = currentFrame.newFrame(isMaxFrameReached());
            frames.add(currentFrame);
        }
        return currentFrame;
    }

    private boolean isMaxFrameReached() {
        return frames.size() >= MAX_FRAMES_PER_GAME;
    }

}

File: C#/BowlingKata/Bowling.cs
using System.Collections.Generic;
using System.Linq;

namespace BowlingKata
{
    public class Bowling
    {
        private static readonly int MaxNumberOfRoll = 10;
        private Frame currentFrame;

        public int ComputeScore(string line)
        {
            var score = 0;
            //ConvertLineToFrame(line).ForEach(frame =>
            //{
            //    score += frame.Pins;
            //});
            score = ConvertLineToFrame(line)
                .Where(frame => !frame.IsBonusFrame())
                .Select(frame => frame.GetFrameScore())
                .Sum();
            return score;

        }

        private List<Frame> ConvertLineToFrame(string line)
        {
            List<Frame> frames = new List<Frame>();

            foreach (var roll in line)
            {
                CurrentFrame(frames).AddRoll(roll.ToString());
            }
            return frames;
        }

        private Frame CurrentFrame(List<Frame> frames)
        {
            if (currentFrame == null)
            {
                currentFrame = new Frame(false);
                frames.Add(currentFrame);
            }
            else if (currentFrame.IsFinished)
            {
                currentFrame = currentFrame.NewFrame(frames.Count >= MaxNumberOfRoll);
                frames.Add(currentFrame);
            }
            return currentFrame;

        }
    }
}
